### PR TITLE
bug: Fix parsing of edge cases phpdoc types

### PR DESCRIPTION
**Pull request by mvorisek**
heavily inspired by phpstan phpdoc grammar and its test cases

### CODE REVIEW

**Comment by mvorisek** [comment]
PR is done, with #6796, #6798, #6800, #6961, #6962 it fixes parsing of almost every valid phpdoc.

I hope a new minor version can be released after this PR soon. Thank you!

File: tests/Fixer/ClassNotation/OrderedTypesFixerTest.php [hunk 1]
```diff
@@ -660,7 +660,7 @@ public function testFixPhp82(string $expected, ?string $input = null, ?array $co
     }
 
     /**
-     * @return iterable<string[]|(string|null|array<string, string>)[]>
+     * @return iterable<(null|array<string, string>|string)[]|string[]>
```
Comment (Wirone) [inline tests/Fixer/ClassNotation/OrderedTypesFixerTest.php:663]: Isn't actually:

```suggestion
     * @return iterable<array<null|string|array<string, string>>>
```

??

I know it's not a part of this PR, but looking at it, seems overcomplicated. Not a something you should change, I would want to discuss it and create separate PR if needed.

File: src/DocBlock/TypeExpression.php [hunk 2]
```diff
@@ -78,24 +83,30 @@ final class TypeExpression
...
-                    (?&name)(\[\])+
-                )
-                |
-                (?<constant> # single constant value (case insensitive), e.g.: 1, `\'a\'`
+                (?<constant> # single constant value (case insensitive), e.g.: 1, -1.8E+6, `\'a\'`
                     (?i)
                     null | true | false
-                    | -?(?:\d+(?:\.\d*)?|\.\d+) # all sorts of numbers with or without minus, e.g.: 1, 1.1, 1., .1, -1
-                    | \'(?:[^\'\\\\]|\\\\.)*?\' | "(?:[^"\\\\]|\\\\.)*?"
+                    | [+-]?(?:
```
Comment (Wirone) [inline src/DocBlock/TypeExpression.php:92]: Not tested for syntax correctness, but I would like to keep comment which was removed.

```suggestion
                    # all sorts of numbers: with or without sign / literal separator, in several numeric system,
                    # e.g.: 1, +1.1, 1., .1, -1, 123E+8, 123_456_789, 0x7Fb4, 0b0110, 0o777
                    | [+-]?(?:
```

**Review by Wirone — CHANGES_REQUESTED** [review]

File: tests/Fixer/ClassNotation/OrderedTypesFixerTest.php (hunk 1 above)
Comment (mvorisek) [reply tests/Fixer/ClassNotation/OrderedTypesFixerTest.php:663]: The best phpdoc here would be using array shape, ie. `iterable<array{ string, null|string, null|array<string, string> }>`.

File: src/DocBlock/TypeExpression.php (hunk 2 above)
Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:92]: I had to revert, your regex comment has broken the regex, feel free to correct it and push to my PR branch

Comment (Wirone) [reply src/DocBlock/TypeExpression.php:92]: That's why I explicitly stated that regex was not tested 😉 OK, let me checkout your branch.

Comment (Wirone) [reply src/DocBlock/TypeExpression.php:94]: Funny, `sign / literal separator` was causing the problem because `/l` was treated as modifier 😅. I've improved the comment and tested `PhpCsFixer\Tests\AutoReview\DocumentationTest` locally, it worked. I hope other tests too 😉.

Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:94]: 👍modifiers are parsed before comments are stripped

**Review by Wirone — APPROVED** [review]
@kubawerlos please take a look too 🙂.
